kasa/auth.py

```diff
@@ -0,0 +1,45 @@
+"""Authentication class for username / passwords."""
+from typing import Dict, Optional
+
+from kasa.exceptions import SmartDeviceException
+from kasa.protocol import TPLinkProtocol
+
+
+class AuthCredentials:
+    """Authentication credentials for Kasa authentication."""
+
+    def __init__(self, username: str = "", password: str = ""):
+        self.username = username
+        self.password = password
+
+
+class TPLinkAuthProtocol(TPLinkProtocol):
+    """Base class for authenticating protocol."""
+
+    def __init__(
+        self,
+        host: str,
+        port: int = 0,
+        auth_credentials: Optional[AuthCredentials] = AuthCredentials(),
+    ):
+        super().__init__(host=host, port=port)
+
+        self.auth_credentials: AuthCredentials
+        if auth_credentials is None:
+            self.auth_credentials = AuthCredentials()
+        else:
+            self.auth_credentials = auth_credentials
+        self._authentication_failed = False
+
+    @property
+    def authentication_failed(self):
+        """Will be true if authentication negotiated but failed, false otherwise."""
+        return self._authentication_failed
+
+    @authentication_failed.setter
+    def authentication_failed(self, value):
+        self._authentication_failed = value
+
+    def parse_unauthenticated_info(self, unauthenticated_info) -> Dict[str, str]:
+        """Return parsed unauthenticated info for the protocol."""
+        raise SmartDeviceException("parse_unauthenticated_info should be overridden")
```

Comment:
I added this for two reasons, firstly kasa's statement that they would be adding a new strong local api to all devices at some point in the future https://community.tp-link.com/en/smart-home/forum/topic/239364?page=1 and secondly so that discovery logic would be "Authentication" aware rather than klap aware.  Happy to remove if you think it's unnecessary

Reply:
This is now removed from the PR